File: c/knights_tour.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#define BOARD_SIZE 5
#define STEPS_SIZE 8

bool solve_knights_tour(int board[BOARD_SIZE][BOARD_SIZE], int steps[STEPS_SIZE][2], int x, int y, int step_count);
bool is_safe(int board[BOARD_SIZE][BOARD_SIZE], int x, int y);
void print_solution(int board[BOARD_SIZE][BOARD_SIZE]);


int main()
{
    // steps is the possible steps for knight at any position. For example {-1, 2} is one left and 2 down, {2, 1} is 2 right and 1 down.
    int steps[STEPS_SIZE][2] = {{-1, 2}, {1, 2}, {-2, 1}, {2, 1}, {1, -2}, {-1, -2}, {2, -1}, {-2, -1}};

    int board[BOARD_SIZE][BOARD_SIZE];
    memset(board, 0, sizeof(board));

    board[0][0] = 1;
    if (solve_knights_tour(board, steps, 0, 0, 2)) {
        printf("Found a solution\n");
        print_solution(board);
    }
    else
        printf("Could not find a solution\n");
    return 0;
}

bool solve_knights_tour(int board[BOARD_SIZE][BOARD_SIZE], int steps[STEPS_SIZE][2], int x, int y, int step_count)
{
    if (step_count > BOARD_SIZE * BOARD_SIZE)
        return true;

    for (int i = 0; i < STEPS_SIZE; i++) {
        int next_x = x + steps[i][0];
        int next_y = y + steps[i][1];

        if (is_safe(board, next_x, next_y)) {
            board[next_x][next_y] = step_count;
            if (solve_knights_tour(board, steps, next_x, next_y, step_count + 1))
                return true;
            board[x][y] = 0;		// Triggers backtracking
        }
    }

    return false;
}

bool is_safe(int board[BOARD_SIZE][BOARD_SIZE], int x, int y)
{
    return x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE && board[x][y] == 0;
}

void print_solution(int board[BOARD_SIZE][BOARD_SIZE])
{
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++)
            board[i][j] < 10 ? printf("0%d ", board[i][j]) : printf("%d ", board[i][j]);
        printf("\n");
    }
}

File: python/knights_tour.py
BOARD_SIZE = 5
board = [[0 for i in range(BOARD_SIZE)] for j in range(BOARD_SIZE)]     # Create a board with size BOARD_SIZE x BOARD_SIZE filled with 0

# STEPS is the possible steps for knight at any position.For example {-1, 2} is one left and 2 down, {2, 1} is 2 right and 1 down
STEPS = [[-1, 2], [1, 2], [-2, 1], [2, 1], [1, -2], [-1, -2], [2, -1], [-2, -1]]


def solve_knights_tour(x, y, step_count):
    if step_count > BOARD_SIZE * BOARD_SIZE:
        return True
    for step in STEPS:
        next_x = x + step[0]
        next_y = y + step[1]
        if is_safe(next_x, next_y):
            board[next_x][next_y] = step_count
            if solve_knights_tour(next_x, next_y, step_count + 1):
                return True
            board[next_x][next_y] = 0       # triggers backtracking
    return False


def is_safe(x, y):
    return 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE and board[x][y] == 0


def print_solution():
    for row in board:
        for col in row:
            print("0" + str(col) if col < 10 else col, end=" ")
        print()


board[0][0] = 1     # First move is at (0, 0)

if solve_knights_tour(0, 0, 2):
    print("Found a solution")
    print_solution()
else:
    print("Could not find a solution")